Add ENGAGE-HF Allergy Form

NikolaiMadlener opened the pull request:
# Add ENGAGE-HF Allergy Form, Observation Form & PatientForm

## :gear: Release Notes 

- Add AllergyForm from ENGAGE-HF and supporting MedicationSelect component 
- Replace custom models used in ENGAGE AllergyForm with [medplum FHIR types](https://www.npmjs.com/package/@medplum/fhirtypes).
- Add ObservationForm from ENGAGE-HF
- Add PatientsTable component
- Update GitHub actions to configure publishing a package
- Remove old Next-related dependencies, files, configurations

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

NikolaiMadlener committed:
add allergy form and medication selection components from engage

NikolaiMadlener committed:
use medplum fhir types

NikolaiMadlener committed:
add dist to gitignore

NikolaiMadlener committed:
remove dist folder

NikolaiMadlener committed:
adjust workflows

NikolaiMadlener committed:
fix invalid hook call error

NikolaiMadlener committed:
add FHIR validation schemas, add allergy and observation FHIR types, add ObservationForm, add PatientForm

NikolaiMadlener committed:
fix license headers

NikolaiMadlener committed:
add UI tests to components

NikolaiMadlener committed:
add non-local spezi web design system package

NikolaiMadlener committed:
update imports

NikolaiMadlener committed:
formatting imports, fix tests by adding server deps inline to vite config

NikolaiMadlener committed:
add missing input test in allergy form

ThomasKaar commented on src/components/AllergyForm/MedicationSelect.tsx:
```diff
@@ -0,0 +1,54 @@
+              : Object.values(medicationClass.name)[0]}
+          </SelectLabel>
+          {medicationClass.medications.map((medication) => (
+            <SelectItem value={medication.code?.coding?.[0].code ?? ""} key={medication.code?.coding?.[0].code ?? ""}>
```
```suggestion
            <SelectItem value={medication.code?.coding?.[0]?.code ?? ""} key={medication.code?.coding?.[0]?.code ?? ""}>
```

ThomasKaar commented on src/modules/fhir/allergy-intolerance.ts:
We are sometimes using `"` and sometimes `'` for string literals
I think this should be just handled by prettier

ThomasKaar reviewed (approved):
Looks great! Thanks a lot for adding some health components! 🚀

Especially, I think it makes a lot of sense to extend the test cases for some components and maybe also add storybook stories so it is easier to inspect the different components and decide whether it is useful for their project, but this would be for separate PRs.

arkadiuszbachorski commented:
I'll take a look at this soon, hopefully tomorrow.

NikolaiMadlener commented on src/components/AllergyForm/AllergyForm.tsx:
Created #4 for this 👍

NikolaiMadlener committed:
Update src/components/AllergyForm/MedicationSelect.tsx

Co-authored-by: ThomasKaar <31984130+ThomasKaar@users.noreply.github.com>

NikolaiMadlener committed:
improve imports and decrease bundle size; remove unused storybook dependencies, ignore config files in codecov

NikolaiMadlener committed:
merge origin

NikolaiMadlener committed:
fix dropdown menu

NikolaiMadlener committed:
lint

NikolaiMadlener committed:
replace conditional with switch case in getObservationTypeUnits

NikolaiMadlener committed:
remove coding value of select in allergy form

NikolaiMadlener committed:
ignore all config files in codecov

NikolaiMadlener committed:
update readme

NikolaiMadlener committed:
fix deployment badge

NikolaiMadlener committed:
fix readme codecov link

arkadiuszbachorski commented on src/index.ts:
```diff
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier: MIT
+//
+
+export * from './components/AllergyForm/AllergyForm'
+export * from './components/PatientForm/PatientForm'
+export * from './components/ObservationForm/ObservationForm'
+export * from './components/PatientsTable/PatientsTable'
```
It seems reasonable to export complete modules

```suggestion
export * from './components/AllergyForm'
export * from './components/PatientForm'
export * from './components/ObservationForm'
export * from './components/PatientsTable'
```

arkadiuszbachorski commented on src/modules/fhir/observation.ts:
```diff
@@ -0,0 +1,62 @@
+  amended = 'amended',
+  corrected = 'corrected',
+  cancelled = 'cancelled',
+  entered_in_error = 'entered-in-error',
```
```suggestion
  enteredInError = 'entered-in-error',
```

arkadiuszbachorski commented on src/modules/fhir/observation.ts:
```diff
@@ -0,0 +1,62 @@
+
+import { z } from 'zod'
+
+export const FHIR_OBSERVATION_RESOURCE_TYPE = 'Observation'
```
Would it make sense to keep enum of FHIR resource types, instead of multiple constants?

NikolaiMadlener replied:
Makes sense; added it to `base.ts` for now.

arkadiuszbachorski commented on src/modules/fhir/base.ts:
```diff
@@ -0,0 +1,32 @@
+
+import { z } from 'zod'
+
+export const FHIRCodableConceptValidationSchema = z.object({
```
I wouldn't be surprised if `@stanfordbdhg/engagehf-models` already contained this validator. Not sure though.

NikolaiMadlener replied:
Indeed, thx for pointing out!

arkadiuszbachorski commented on src/modules/fhir/base.ts:
```diff
@@ -0,0 +1,32 @@
+
+import { z } from 'zod'
+
+export const FHIRCodableConceptValidationSchema = z.object({
```
Wouldn't `Schema` postfix instead of `ValidationSchema` be enough?

arkadiuszbachorski commented on src/modules/fhir/allergy-intolerance.ts:
```diff
@@ -0,0 +1,113 @@
+]
+
+export const MANIFESTATION_OPTIONS = [
+  { code: '41291007', display: 'Angioedema' },
+  { code: '271807003', display: 'Rash' },
+  { code: '49727002', display: 'Cough' },
+  { code: '74964007', display: 'Other' },
```
Could these codes be explained with an enum?

arkadiuszbachorski commented on src/components/PatientsTable/PatientsTable.tsx:
```diff
@@ -0,0 +1,59 @@
+  editRoute,
+  ...props
+}: PatientsDataTableProps) => {
+  const columnHelper = createColumnHelper<Patient>()
```
It can be computed outside of `PatientsTable` render body, as it has no props/state references. This way, it's just computed once.

arkadiuszbachorski commented on src/components/PatientsTable/PatientsTable.tsx:
```diff
@@ -0,0 +1,59 @@
+  const columnHelper = createColumnHelper<Patient>()
+
+  const columns = [
+    { header: 'First Name', accessorKey: 'name.0.given' },
+    { header: 'Last Name', accessorKey: 'name.0.family' },
```
Let's use `columnHelper` for all the columns if possible. This way, we get full type-safety.

Another point - `accessorKey` as string are not worth it. With TypeScript, it's better to just write an accessor function, because we get all the type hints. I am aware that `accessKey` most likely also use type checking to verify this path exists, but from my experience those create problems sooner or later.

arkadiuszbachorski commented on src/components/PatientsTable/PatientsTable.tsx:
```diff
@@ -0,0 +1,59 @@
+
+interface PatientsDataTableProps
+  extends Omit<DataTableProps<Patient>, 'columns'> {
+  onRowClick?: (data: Patient) => void
```
Let's keep standard `DataTable` interface, therefore, `onRowClick` passed through `tableView` object.  Less code, same API structure, less mental overhead mapping them.

`tableView` exist as a separate prop for a reason. `DataTable` can be rendered with completely custom view using `children`. `tableView` refers to table-view specific props.

arkadiuszbachorski commented on src/components/PatientsTable/PatientsTable.tsx:
```diff
@@ -0,0 +1,59 @@
+  extends Omit<DataTableProps<Patient>, 'columns'> {
+  onRowClick?: (data: Patient) => void
+  onDelete: (data: Patient) => void
+  editRoute: string
```
I think it would make sense to create some health-components-wide context that gathers all the routes. Routes might be different per application usage, but they're constant. It seems like it makes sense to inject them in just one place and consume that in the components.

arkadiuszbachorski replied:
`editRoute` cannot be a plain `string`. Edit routes usually contain ID, so it should be:

```
(patientId: string) => string
```

arkadiuszbachorski commented on src/components/PatientsTable/PatientsTable.tsx:
```diff
@@ -0,0 +1,59 @@
+import { createColumnHelper } from '@tanstack/table-core'
+import { PatientMenu } from './PatientMenu'
+
+interface PatientsDataTableProps
```
To match `<ComponentName>Props` naming pattern.

```suggestion
interface PatientsTableProps
```

arkadiuszbachorski commented on src/components/PatientsTable/PatientsTable.tsx:
```diff
@@ -0,0 +1,59 @@
+interface PatientsDataTableProps
+  extends Omit<DataTableProps<Patient>, 'columns'> {
+  onRowClick?: (data: Patient) => void
+  onDelete: (data: Patient) => void
```
```suggestion
  onDelete: (patient: Patient) => void
```

arkadiuszbachorski commented on src/components/PatientsTable/PatientsTable.tsx:
Do we treat this component as just a starting point that could be referenced to modify? Columns are possible too vague to serve most projects. I'd also consider adding `email` to columns if `Patient` contain this information.

NikolaiMadlener replied:
I would see it as a starting point for potential future improvements. Probably a lot of applications will have some patient table and yes, columns might be too vague but still, if one looks for a fast way to list patients, this might be enough.

arkadiuszbachorski commented on src/components/PatientsTable/PatientsTable.test.tsx:
```diff
@@ -0,0 +1,39 @@
+  it('renders element', () => {
+    const onDelete = vitest.fn()
+    const editRoute = '/patients/edit'
+    const data: Patient[] = [
+      { id: '1', name: [{ given: ['Leland'], family: 'Stanford' }] } as Patient,
+    ]
```
`Patient[]` is unnecessary. TypeScript already knows it's `Patient[]` due to `as Patient` declaration.

arkadiuszbachorski commented on src/components/PatientsTable/PatientsTable.test.tsx:
```diff
@@ -0,0 +1,39 @@
+describe('PatientsTable', () => {
+  it('renders element', () => {
+    const onDelete = vitest.fn()
+    const editRoute = '/patients/edit'
```
This can be just declared inline. If it's not reference anywhere, it doesn't have to be complicated

arkadiuszbachorski commented on src/components/PatientsTable/PatientsTable.test.tsx:
```diff
@@ -0,0 +1,39 @@
+      <PatientsTable onDelete={onDelete} editRoute={editRoute} data={data} />,
+    )
+
+    const firstNameHeader = screen.getByText('First Name', {
+      selector: 'button',
+    })
+    expect(firstNameHeader).toBeInTheDocument()
+
+    const lastNameHeader = screen.getByText('Last Name', { selector: 'button' })
+    expect(lastNameHeader).toBeInTheDocument()
+
+    const firstNameCell = screen.getByText('Leland', { selector: 'td' })
+    expect(firstNameCell).toBeInTheDocument()
+
+    const lastNameCell = screen.getByText('Stanford', { selector: 'td' })
+    expect(lastNameCell).toBeInTheDocument()
```
I think selectors here do not make much sense. It's an implementation detail of `DataTable` component, it's tested within `DataTable`.

arkadiuszbachorski commented on src/components/PatientsTable/PatientMenu.tsx:
```diff
@@ -0,0 +1,57 @@
+import { Link } from '@tanstack/react-router'
+import { Pencil, Trash } from 'lucide-react'
+
+interface PatientMenuProps {
+  onDelete: (data: Patient) => void
+  editRoute: string
+  data: Patient
+}
```
If it's already in "patient" context, it seems to make sense to adjust names to reflect what data is passed.

```suggestion
interface PatientMenuProps {
  onDelete: (patient: Patient) => void
  editRoute: string
  patient: Patient
}
```

arkadiuszbachorski commented on src/components/PatientForm/PatientForm.tsx:
```diff
@@ -0,0 +1,67 @@
+export const PatientForm = ({ patient, onSubmit }: PatientFormProps) => {
+  const isEdit = !!patient
+  const form = useForm({
+    formSchema: patientFormSchema,
```
`defaultValue` needs to be provided for edit variant.

arkadiuszbachorski commented on src/components/PatientForm/PatientForm.test.tsx:
```diff
@@ -0,0 +1,27 @@
+import { PatientForm } from '.'
+
+describe('PatientForm', () => {
+  it('renders element', () => {
```
Would be great to test submission structures and form behavior

arkadiuszbachorski commented on src/components/ObservationForm/ObservationForm.tsx:
```diff
@@ -0,0 +1,205 @@
+          <Input
+            {...field}
+            type="number"
+            onChange={(event) =>
+              field.onChange(event.currentTarget.valueAsNumber)
```
This is unnecessary. `useForm` can handle numbers right away, without overriding `onChange`.

NikolaiMadlener replied:
Removing the explicit cast, the form gives me this when trying to submit...
![image](https://github.com/user-attachments/assets/da01d33d-41de-40f0-8484-39ffa3f208ff)

arkadiuszbachorski replied:
Yes, sorry, that's correct. 

`useForm` is able to cast correct types, but only if validation schema does it. In this case `z.number()` expects actual `Number` type. In order to have casting, we'd have to use `z.coerce.number()`, but let's keep as is.

arkadiuszbachorski reviewed (changes requested):
Thank you for working on that! I left my points. Some of them are nits, some can be addressed in separate PRs. Please, create issues if handled as separate. 

Some comments might require to be handled PR-wise. I didn't create a comment for every instance where issue exists, like `Array.at` thing.

I didn't finish my review, but that's what I have so far. Please let me know if any questions.

NikolaiMadlener committed:
Update .github/workflows/deployment.yml

Co-authored-by: Arkadiusz Bachorski <60391032+arkadiuszbachorski@users.noreply.github.com>

NikolaiMadlener committed:
Update package.json

Co-authored-by: Arkadiusz Bachorski <60391032+arkadiuszbachorski@users.noreply.github.com>

NikolaiMadlener committed:
clean up package.json and remove redundant dependencies

NikolaiMadlener committed:
Update src/components/AllergyForm/AllergyForm.test.tsx

Co-authored-by: Arkadiusz Bachorski <60391032+arkadiuszbachorski@users.noreply.github.com>

NikolaiMadlener committed:
resolve merge conflict

NikolaiMadlener commented on src/components/AllergyForm/AllergyForm.test.tsx:
```diff
@@ -0,0 +1,57 @@
+      <AllergyForm medicationClasses={medicationClasses} onSubmit={onSubmit} />,
+    )
+
+    const allergyTypeSelector = screen.getByText('Type', { selector: 'label' })
```
Thanks for the hint! 
In iOS-land I would probably call it Picker to Selector, yes.

NikolaiMadlener commented on src/modules/fhir/allergy-intolerance.ts:
```diff
@@ -0,0 +1,113 @@
+}
+
+export enum FHIRAllergyClinicalStatus {
+  // FHIR clinical status following clinical status coding system
```
pretty redundant comment as well, removed it.

NikolaiMadlener commented on src/components/AllergyForm/AllergyForm.tsx:
```diff
@@ -0,0 +1,201 @@
+  allergy?: AllergyIntolerance
+  medicationClasses: Array<{
+    id: string
+    name: string | Record<string, string>
```
Moved the type definition to a new LocalizedText file;  lmk what you think.

NikolaiMadlener commented on src/components/AllergyForm/AllergyForm.tsx:
```diff
@@ -0,0 +1,201 @@
+      }
+      await onSubmit(allergyIntolerance)
+    } catch (error) {
+      console.error('Form submission error:', error)
```
Agree; created a ticket to tackle it in the future.

NikolaiMadlener commented on src/components/AllergyForm/MedicationSelect.tsx:
```diff
@@ -0,0 +1,57 @@
+      {medicationClasses.map((medicationClass) => (
+        <SelectGroup key={medicationClass.id}>
+          <SelectLabel>
+            {typeof medicationClass.name === 'string' ?
+              medicationClass.name
+            : Object.values(medicationClass.name)[0]}
```
Copied some of the code over from engage. Once a first version of the package is released we could think of using the `parseLocalizedText` directly from web-health-components within engage.

NikolaiMadlener commented on src/components/AllergyForm/AllergyForm.tsx:
```diff
@@ -0,0 +1,201 @@
+      />
+      <Button
+        type="submit"
+        disabled={!form.formState.isValid}
```
Good point!

NikolaiMadlener commented on src/components/AllergyForm/AllergyForm.tsx:
```diff
@@ -0,0 +1,201 @@
+        label="Type"
+        render={({ field }) => (
+          <Select onValueChange={(value) => field.onChange(value)} {...field}>
+            <SelectTrigger id="type">
```
When I remove the `id` from `SelectTrigger`, my tests start failing. Seems like the SelectTrigger needs to have the id of the label to be associated with the label. 
Any idea?

arkadiuszbachorski replied:
I investigated that. 

Apparently, `Select` component does not propagate it's id to `SelectTrigger`, so the `id` from `field` is gone. 

This breaks tests, because there is no form element associated with the `id` label.

I'll update `Select` component to actually follow this pattern and forward `id` to `SelectTrigger` if it exists. That's one of the principles of `Field`, to make labeling and error to happen implicitly.

arkadiuszbachorski replied:
https://github.com/StanfordSpezi/spezi-web-design-system/issues/33
https://github.com/StanfordSpezi/spezi-web-health-components/issues/10

NikolaiMadlener committed:
implement PR feedback

NikolaiMadlener commented:
@arkadiuszbachorski Thanks for the detailed feedback, extremely helpful!
Most of the points should be resolved now (with some being moved into new Github issues). 
Feel free to resolve the discussions above if the adjustments look good to you. 

Excited to get this merged soon 🚀

NikolaiMadlener commented on src/components/AllergyForm/AllergyForm.test.tsx:
```diff
@@ -0,0 +1,57 @@
+import { AllergyForm } from '.'
+
+describe('AllergyForm', () => {
+  it('renders element', () => {
```
Totally agree! Tried around a bit but had some issues getting the tests passing. Might be nice to have a look at it together at some point. For now, I pushed it without the new tests. Could also be addressed in a future PR (created an issue as well)

arkadiuszbachorski commented on src/components/AllergyForm/AllergyForm.tsx:
```diff
@@ -0,0 +1,210 @@
+  })
+
+  const handleSubmit = form.handleSubmit(async (data) => {
+    console.log('Form values before submission:', data)
```
Leftover

arkadiuszbachorski committed:
Updates to Allergy From (#11)

# Updates to Allergy From

This mostly concludes missing elements from my feedback. There are also
other missing elements, but I'll create separate issues/PRs for that. I
just want to merge what is already there, to simplify the process.

Feel free to ask if anything is not clear enough.

arkadiuszbachorski reviewed (approved):
I just merged my PR, let's ship this 🚀
